vault/auth.go

```diff
@@ -820,6 +820,13 @@ func (c *Core) setupCredentials(ctx context.Context) error {
 			// don't set the running version to a builtin if it is running as an external plugin
 			if externaler, ok := backend.(logical.Externaler); !ok || !externaler.IsExternal() {
 				entry.RunningVersion = versions.GetBuiltinVersion(consts.PluginTypeCredential, entry.Type)
+				if _, err := c.handleDeprecatedMountEntry(ctx, entry, consts.PluginTypeCredential); err != nil {
+					c.logger.Error("shutting down core", "error", err)
+					if shutdownErr := c.Shutdown(); shutdownErr != nil {
```

Comment:
Why do we shutdown here? It feels quite risky to me as it's going to be a very difficult situation for a user to recover from, and wouldn't returning an error achieve the same result in terms of stopping Vault from creating the mount/starting up?

Instead of failing though, is it possible to just not start a plugin, but leave the mount data in tact and log the error? Then, users can decide whether they want to try and recover the mount and/or its data by tuning it to use an external version, or just delete the mount.

Reply:
This is meant to handle the upgrade case. Once we find a way to block deregistering mounted plugins, this should be the safest route for an upgrade, as it would prevent future divergence from the pre-upgrade state.

Reply:
@tomhjp when you have some time, I've attempted to resolve this by introducing a [new entry](https://github.com/hashicorp/vault/pull/17879/commits/d5a5e45aa88f5558c9dd8032c24d4052b8eb7776) to the version store, which keeps track of the last successful mount. This should only be recorded after all secrets engines and auth methods are successfully mounted. Combined with blocking of deregisters, this should provide the desired behavior of:

1) Force shutdown on first unseal post-upgrade if there are deprecated mounts detected
2) Lazy mount if builtins are deregistered post-upgrade.

Where the latter is just a safety net and should never occur if we're blocking deregistration, etc. (caveat: replication considerations, which are out of scope for this PR).

Reply:
I decided to use @ncabatoff's suggested approach and instead move this logic into the MountEntries, indicating the last version of Vault for which a mount was successfuly as `LastMounted`. This is a much cleaner and potentially more useful approach.